### PR TITLE
[CodeGen] Avoid potential sideeffects from XOR

### PR DESCRIPTION
**Pull request by bwendling**
XOR may change flag values (e.g. for X86 gprs). In the case where that's not desirable, specify that buildClearRegister() should use MOV instead.

### CODE REVIEW

**Comment by RKSimon** [comment]
The new flag doesn't appear to be used yet - are additional patches that wire it up and add tests?

**Comment by bwendling** [comment]
@RKSimon Yes. I plan on using it to clear out registers that have the stack protector guard value in them before exit.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 1]
```diff
@@ -9066,21 +9066,18 @@ bool AArch64InstrInfo::shouldOutlineFromFunctionByDefault(
...
+    BuildMI(MBB, Iter, DL, get(AArch64::MOVZXi), Reg).addImm(0).addImm(0);
   } else if (STI.hasSVE()) {
-    BuildMI(MBB, Iter, DL, get(AArch64::DUP_ZI_D), Reg)
-      .addImm(0)
-      .addImm(0);
+    BuildMI(MBB, Iter, DL, get(AArch64::DUP_ZI_D), Reg).addImm(0).addImm(0);
   } else {
-    BuildMI(MBB, Iter, DL, get(AArch64::MOVIv2d_ns), Reg)
-      .addImm(0);
+    BuildMI(MBB, Iter, DL, get(AArch64::MOVIv2d_ns), Reg).addImm(0);
```
Comment (RKSimon) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:9080]: NFC clang-format change - remove this from the patch

File: llvm/lib/Target/X86/X86InstrInfo.h [hunk 2]
```diff
@@ -574,8 +574,8 @@ class X86InstrInfo final : public X86GenInstrInfo {
                      outliner::Candidate &C) const override;
 
   void buildClearRegister(Register Reg, MachineBasicBlock &MBB,
-                          MachineBasicBlock::iterator Iter,
-                          DebugLoc &DL) const override;
+                          MachineBasicBlock::iterator Iter, DebugLoc &DL,
+                          bool NoSideEffects = false) const override;
```
Comment (nickdesaulniers) [inline llvm/lib/Target/X86/X86InstrInfo.h:578]: who calls buildClearRegister with this new default param set to true?

also, I fear double negatives here. Would it be clearer to call this `SideEffects` and default to `true`?

File: llvm/lib/Target/X86/X86InstrInfo.cpp [hunk 3]
```diff
@@ -9796,27 +9796,34 @@ X86InstrInfo::insertOutlinedCall(Module &M, MachineBasicBlock &MBB,
...
   if (ST.hasMMX() && X86::VR64RegClass.contains(Reg))
-    // FIXME: Ignore MMX registers?
+    // FIXME: Should we ignore MMX registers?
     return;
 
   if (TRI.isGeneralPurposeRegister(MF, Reg)) {
-    BuildMI(MBB, Iter, DL, get(X86::XOR32rr), Reg)
-      .addReg(Reg, RegState::Undef)
-      .addReg(Reg, RegState::Undef);
+    // Convert register to the 32-bit version.
```
Comment (nickdesaulniers) [inline llvm/lib/Target/X86/X86InstrInfo.cpp:9811]: Maybe note that xor32rr and mov32ri will clear the upper 32b (if that's the case?)

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 4]
```diff
@@ -9066,21 +9066,18 @@ bool AArch64InstrInfo::shouldOutlineFromFunctionByDefault(
...
+                                          DebugLoc &DL,
+                                          bool NoSideEffects) const {
   const MachineFunction &MF = *MBB.getParent();
   const AArch64Subtarget &STI = MF.getSubtarget<AArch64Subtarget>();
   const AArch64RegisterInfo &TRI = *STI.getRegisterInfo();
 
   if (TRI.isGeneralPurposeRegister(MF, Reg)) {
-    BuildMI(MBB, Iter, DL, get(AArch64::MOVi64imm), Reg)
-      .addImm(0);
+    BuildMI(MBB, Iter, DL, get(AArch64::MOVZXi), Reg).addImm(0).addImm(0);
```
[the remark this answers is not in the document]
Comment (bwendling) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp]: Yes. This is how it's done for other uses of `MOVZXi`. I'll verify that it's required though.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp (hunk 1 above)
Comment (bwendling) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp]: Okay.

File: llvm/lib/Target/X86/X86InstrInfo.cpp (hunk 3 above)
Comment (bwendling) [reply llvm/lib/Target/X86/X86InstrInfo.cpp]: Done.

File: llvm/lib/Target/X86/X86InstrInfo.h (hunk 2 above)
Comment (bwendling) [reply llvm/lib/Target/X86/X86InstrInfo.h:578]: I'm going to be using this for clearing registers using stack protectors. I don't think it's clearer to use `SideEffects`. I know it's a double negative, but it just feels more natural to me to say "I want the instruction to have no side effects."

Comment (bwendling) [reply llvm/lib/Target/X86/X86InstrInfo.h:578]: Perhaps it should be `AvoidSideEffects`? Then it wouldn't not haven't had that double negative...

Comment (nickdesaulniers) [reply llvm/lib/Target/X86/X86InstrInfo.h]: Yeah, that SGTM

**Review by nickdesaulniers — APPROVED** [review]

File: llvm/include/llvm/CodeGen/TargetInstrInfo.h [hunk 5]
```diff
@@ -2088,12 +2088,12 @@ class TargetInstrInfo : public MCInstrInfo {
   }
 
   /// Insert an architecture-specific instruction to clear a register. If you
-  /// need to avoid sideeffects (e.g. XOR on x86), set \p NoSideEffects to \p
-  /// true.
+  /// need to avoid sideeffects (e.g. XOR on x86), set \p AllowSideEffects to
```
Comment (RKSimon) [inline llvm/include/llvm/CodeGen/TargetInstrInfo.h:2091]: (e.g. avoid XOR on x86 which sets EFLAGS)

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 6]
```diff
@@ -9066,21 +9066,23 @@ bool AArch64InstrInfo::shouldOutlineFromFunctionByDefault(
...
   } else if (STI.hasSVE()) {
     BuildMI(MBB, Iter, DL, get(AArch64::DUP_ZI_D), Reg)
-      .addImm(0)
-      .addImm(0);
+        .addImm(0)
+        .addImm(0);
   } else {
     BuildMI(MBB, Iter, DL, get(AArch64::MOVIv2d_ns), Reg)
-      .addImm(0);
+        .addImm(0);
```
Comment (RKSimon) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:9085]: remove these NFC indentation changes

Comment (bwendling) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp]: Done.

File: llvm/include/llvm/CodeGen/TargetInstrInfo.h (hunk 5 above)
Comment (bwendling) [reply llvm/include/llvm/CodeGen/TargetInstrInfo.h]: Done.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (bwendling) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp]: It does appear to be related to the "shift" argument of `MOVZXi`.

**Pull request merged**